### PR TITLE
feat(primitives): accept iterators for ordered trie root calcs

### PR DESCRIPTION
**Pull request by bsh98**
Modifies the ordered trie root calculation functions to accept iterators instead of slices.  This makes fewer assumptions about function inputs, which improves usability and performance.  For example, this could help users avoid unnecessarily allocating when computing trie roots.

### CODE REVIEW

**Comment by bsh98** [comment]
(apologies for all of the clippy/fmt errors, my config was broken :S should be fixed now!)

**Review by mattsse — COMMENTED** [review]
having iterator as arg is much nicer.

not exactly sure about the merge_by, I think this makes sense, but defer to @rkrasiuk

**Review by rkrasiuk — CHANGES_REQUESTED** [review]
1. Iterators should be over references rather than take ownership of the collections
2. Unclear the need for `merge_by`
3. Pls, remove semicolon formatting

**Comment by bsh98** [comment]
Thank you for the review, @rkrasiuk!

> 1. Iterators should be over references rather than take ownership of the collections

Apologies if I'm missing something here-- where exactly is ownership being taken?  Do you mean changing `impl IntoIterator<..>` to `impl Interator<..>`?  Accepting `impl IntoIterator<..>` allows for consuming collections, but it is not required since `Iter<..>` implements `IntoIter<..>`.  One can call `func(my_collection)` or `func(my_collection.iter())` or `func(&my_vec)`, which in my opinion provides better call site flexibility.  I was under the impression that this is a best practice for functions that accept iterators.  However, I do understand if we want to restrict the user to exclusively operate over references.  

In [e7e68a3](https://github.com/paradigmxyz/reth/pull/6341/commits/e7e68a3816ec86b864f62c47726606aa92bfe448) I changed most of the iterator item generics from `AsRef<T>` to `Borrow<T>` so this is a little more clear.

> 2. Unclear the need for `merge_by`

Perhaps there is a more obvious approach, but my thought process was as follows:

The reason for `merge_by` is that when building the ordered trie, we need to make sure the keys are sorted.  Since the keys are the rlp encoding of the index, special care must be taken only  for the 0th index since the rlp encoding of `0x00` is `0x80`.  Previously, this was handled by [`adjust_index_for_rlp`](https://github.com/paradigmxyz/reth/blob/613ef9186451fff5a24c8014b86b734ae9e198ca/crates/primitives/src/proofs.rs#L16), but since we don’t know the length of the iterator, we can’t use direct indexing.  The idea is that all we need to do is delay the insertion of the 0th indexed element to either when it is the last item to be inserted or we are at index `0x7f`, whichever comes first.  This is equivalent to merging two sorted iterators.  `merge_by` isn’t the only way to do this, but it was cleaner than the other alternatives that I thought of.  Do you have suggestions for a different approach?

> 3. Pls, remove semicolon formatting

Sorry about that.  They should be removed now.

**Comment by mattsse** [comment]
@rkrasiuk please advise whether or how you should proceed here

File: crates/primitives/src/proofs.rs [hunk 1]
```diff
@@ -24,63 +25,87 @@ pub const fn adjust_index_for_rlp(i: usize, len: usize) -> usize {
...
-
+        let key = Nibbles::unpack(&index_buffer);
+        (key, item)
+    });
+    let first = remaining.next().into_iter();
+
+    // Since we know all keys will be ordered, except for the key at the 0th index, there is no
+    // need to re-sort the entire iterator.  All we need to do is insert the 0th index item at the
+    // appropriate index (either rlp(0)==0x7f or last).  We can do this by merging two iterators:
+    // one that contains only the first item, and one that contains the remaining items.
```
Comment (rakita) [inline crates/primitives/src/proofs.rs:58]: Is this statement correct? Previously we used `adjust_index_for_rlp` to increment the index.

```
/// Adjust the index of an item for rlp encoding.
pub const fn adjust_index_for_rlp(i: usize, len: usize) -> usize {
    if i > 0x7f {
        i
    } else if i == 0x7f || i + 1 == len {
        0
    } else {
        i + 1
    }
}
```

For more than `0x7f` (127) items index becomes `i` not `i+1`

**Comment by onbjerg** [comment]
I assume this is still relevant, so re-opening. If not, feel free to close 😊

**Pull request closed without merging**